### PR TITLE
Fix `get_handle_path()`

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Motivated on discord: https://discord.com/channels/691052431525675048/742884593551802431/802272336774299648

### CODE REVIEW

**Comment by cart** [comment]
This does add another atomic operation (handle_to_path.write()) to a function that will get called a lot. Probably not _terrible_ but worth calling out.

We could move it into load_async() which would ensure the op happens on a separate thread, but that has the downside of making things like this not work reliably:

```rust
let handle = server.load("path"); 
server.get_handle_path(handle);
```

**Comment by TheRawMeatball** [comment]
Hmm, my first instinct is to get [dashmap](https://crates.io/crates/dashmap) as a dependency - it would probably solve the performance without such a concurrency issue. It only depends on `num_cpus` and `cfg-if`, both of which already in the build tree, so I'd say its worth considering.

**Comment by cart** [comment]
Hmm I'm not familiar enough with dashmap's impl to have an immediate response. Pulling in a new dependency will always require a bit more scrutiny than normal, so it would be good to evaluate a couple of things before rolling with that approach:

1. Is it really a performance improvement and if so, by how much? Imo contention should be pretty uncommon, so a simple benchmark doing a bunch of loads in a loop would resolve most of my concerns.
2. Does it affect compile times? A couple of clean compiles with/without asset server using dashmap would resolve that question.

**Comment by NathanSWard** [comment]
> We could move it into load_async() which would ensure the op happens on a separate thread, but that has the downside of making things like this not work reliably

I actually don't mind this solution of moving it into `load_async`. As I could easily see a scenario where multiple parallel systems are trying to load some assets and get blocked by `.write()` call on the internal hash map. 

It's definitely a bummer that the `get_handle_path` won't work immediately, however it's a tradeoff I'm happy to take.

**Comment by NathanSWard** [comment]
I was pointed to this PR via #2306 and this is a pretty glaringly missing feature.
We definitely need to support `get_handle_path` as it has use cases, and not to mention that the `AssetServer`'s internal `handle_to_path` `HashMap` is currently just wasting memory.

File: crates/bevy_asset/src/asset_server.rs [hunk 1]
```diff
@@ -358,7 +358,10 @@ impl AssetServer {
                 }
             })
             .detach();
-        asset_path.into()
+        let owned_path = asset_path.to_owned();
+        let id: HandleId = asset_path.into();
+        self.server.handle_to_path.write().insert(id, owned_path);
+        id
```
Comment (NathanSWard) [inline crates/bevy_asset/src/asset_server.rs:364]: Just as an addition to this, could we add a test to validate this behavior?

**Review by NathanSWard — CHANGES_REQUESTED** [review]

**Comment by NathanSWard** [comment]
#2268 and #2306 are both blocked by this PR. 
@TheRawMeatball let me know if you want to help finalize this PR (with the last few changes) or I'm also more than happy to open another PR like this which addresses the issues haha ☺️☺️

**Comment by TheRawMeatball** [comment]
Help is always appreciated :) If you want to push this PR to completion, you can open a new PR using this as a base if you want and I can close this one. I've been fairly short on time recently, so it'd probably be better.

**Comment by NathanSWard** [comment]
> Help is always appreciated :) If you want to push this PR to completion, you can open a new PR using this as a base if you want and I can close this one. I've been fairly short on time recently, so it'd probably be better.

No worries! I'll probably end up making a new PR with this as a base :)

**Comment by NathanSWard** [comment]
Closed in favor of #2310

**Pull request closed without merging**